File: java/src/test/java/com/coinbase/x402/model/Erc4337ConstantsTest.java
package com.coinbase.x402.model;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class Erc4337ConstantsTest {

    /* -------- address constants are valid hex -------------------------------- */

    private static final List<String> ALL_ADDRESSES = List.of(
            Erc4337Constants.ENTRY_POINT_07_ADDRESS,
            Erc4337Constants.SAFE_4337_MODULE_ADDRESS,
            Erc4337Constants.SAFE_WEBAUTHN_SHARED_SIGNER,
            Erc4337Constants.FCL_P256_VERIFIER,
            Erc4337Constants.P256_OWNER_FACTORY,
            Erc4337Constants.WEBAUTHN_SIGNER_FACTORY
    );

    @Test
    void allAddressConstantsAreValidHex() {
        for (String address : ALL_ADDRESSES) {
            assertNotNull(address, "Address constant should not be null");
            assertEquals(42, address.length(),
                    "Address should be 42 characters (0x + 40 hex digits): " + address);
            assertTrue(address.startsWith("0x"),
                    "Address should start with 0x: " + address);
            assertTrue(address.substring(2).matches("[0-9a-fA-F]{40}"),
                    "Address should contain only hex digits after 0x: " + address);
        }
    }

    @Test
    void entryPointAddressIsValid() {
        assertEquals(42, Erc4337Constants.ENTRY_POINT_07_ADDRESS.length());
        assertTrue(Erc4337Constants.ENTRY_POINT_07_ADDRESS.startsWith("0x"));
    }

    @Test
    void safeModuleAddressIsValid() {
        assertEquals(42, Erc4337Constants.SAFE_4337_MODULE_ADDRESS.length());
        assertTrue(Erc4337Constants.SAFE_4337_MODULE_ADDRESS.startsWith("0x"));
    }

    @Test
    void webAuthnSharedSignerAddressIsValid() {
        assertEquals(42, Erc4337Constants.SAFE_WEBAUTHN_SHARED_SIGNER.length());
        assertTrue(Erc4337Constants.SAFE_WEBAUTHN_SHARED_SIGNER.startsWith("0x"));
    }

    @Test
    void fclP256VerifierAddressIsValid() {
        assertEquals(42, Erc4337Constants.FCL_P256_VERIFIER.length());
        assertTrue(Erc4337Constants.FCL_P256_VERIFIER.startsWith("0x"));
    }

    @Test
    void p256OwnerFactoryAddressIsValid() {
        assertEquals(42, Erc4337Constants.P256_OWNER_FACTORY.length());
        assertTrue(Erc4337Constants.P256_OWNER_FACTORY.startsWith("0x"));
    }

    @Test
    void webAuthnSignerFactoryAddressIsValid() {
        assertEquals(42, Erc4337Constants.WEBAUTHN_SIGNER_FACTORY.length());
        assertTrue(Erc4337Constants.WEBAUTHN_SIGNER_FACTORY.startsWith("0x"));
    }

    /* -------- AA_ERROR_MESSAGES ---------------------------------------------- */

    @Test
    void aaErrorMessagesIsNonNull() {
        assertNotNull(Erc4337Constants.AA_ERROR_MESSAGES);
    }

    @Test
    void aaErrorMessagesIsUnmodifiable() {
        assertThrows(UnsupportedOperationException.class, () ->
                Erc4337Constants.AA_ERROR_MESSAGES.put("AA99", "should fail"));
    }

    @Test
    void aaErrorMessagesHasExpectedEntryCount() {
        assertEquals(20, Erc4337Constants.AA_ERROR_MESSAGES.size());
    }

    @Test
    void allExpectedAACodesArePresent() {
        List<String> expectedCodes = List.of(
                "AA10", "AA13", "AA14", "AA15",
                "AA20", "AA21", "AA22", "AA23", "AA24", "AA25", "AA26",
                "AA30", "AA31", "AA32", "AA33", "AA34",
                "AA40", "AA41",
                "AA50", "AA51"
        );

        for (String code : expectedCodes) {
            assertTrue(Erc4337Constants.AA_ERROR_MESSAGES.containsKey(code),
                    "AA_ERROR_MESSAGES should contain code: " + code);
            String message = Erc4337Constants.AA_ERROR_MESSAGES.get(code);
            assertNotNull(message, "Message for " + code + " should not be null");
            assertFalse(message.isEmpty(), "Message for " + code + " should not be empty");
        }
    }

    @Test
    void aa10MessageIsCorrect() {
        assertEquals("Sender already constructed", Erc4337Constants.AA_ERROR_MESSAGES.get("AA10"));
    }

    @Test
    void aa21MessageIsCorrect() {
        assertEquals("Insufficient funds for gas prefund", Erc4337Constants.AA_ERROR_MESSAGES.get("AA21"));
    }

    @Test
    void aa24MessageIsCorrect() {
        assertEquals("Signature validation failed", Erc4337Constants.AA_ERROR_MESSAGES.get("AA24"));
    }

    @Test
    void aa51MessageIsCorrect() {
        assertEquals("Prefund below actualGasCost", Erc4337Constants.AA_ERROR_MESSAGES.get("AA51"));
    }

    @Test
    void allMessagesAreNonEmpty() {
        for (Map.Entry<String, String> entry : Erc4337Constants.AA_ERROR_MESSAGES.entrySet()) {
            assertNotNull(entry.getValue(), "Message for " + entry.getKey() + " should not be null");
            assertFalse(entry.getValue().isEmpty(),
                    "Message for " + entry.getKey() + " should not be empty");
        }
    }
}

File: go/mechanisms/evm/erc4337_constants_test.go
package evm

import (
	"strings"
	"testing"
)

func TestAddressConstants_ValidHex(t *testing.T) {
	addresses := map[string]string{
		"EntryPoint07Address":      EntryPoint07Address,
		"Safe4337ModuleAddress":    Safe4337ModuleAddress,
		"SafeWebAuthnSharedSigner": SafeWebAuthnSharedSigner,
		"FCLP256Verifier":          FCLP256Verifier,
		"P256OwnerFactory":         P256OwnerFactory,
		"WebAuthnSignerFactory":    WebAuthnSignerFactory,
	}

	for name, addr := range addresses {
		t.Run(name, func(t *testing.T) {
			// Must start with 0x
			if !strings.HasPrefix(addr, "0x") {
				t.Errorf("%s should start with 0x, got: %s", name, addr)
			}

			// Must be 42 characters (0x + 40 hex chars)
			if len(addr) != 42 {
				t.Errorf("%s length = %d, want 42", name, len(addr))
			}

			// All characters after 0x must be valid hex
			hexPart := addr[2:]
			for _, c := range hexPart {
				if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
					t.Errorf("%s contains invalid hex character: %c", name, c)
				}
			}
		})
	}
}

func TestAAErrorMessages_ContainsExpectedCodes(t *testing.T) {
	expectedCodes := []string{
		"AA10", "AA13", "AA14", "AA15",
		"AA20", "AA21", "AA22", "AA23", "AA24", "AA25", "AA26",
		"AA30", "AA31", "AA32", "AA33", "AA34",
		"AA40", "AA41",
		"AA50", "AA51",
	}

	for _, code := range expectedCodes {
		t.Run(code, func(t *testing.T) {
			msg, ok := AAErrorMessages[code]
			if !ok {
				t.Errorf("AAErrorMessages missing code %s", code)
				return
			}
			if msg == "" {
				t.Errorf("AAErrorMessages[%s] is empty", code)
			}
		})
	}
}

func TestAAErrorMessages_AllNonEmpty(t *testing.T) {
	for code, msg := range AAErrorMessages {
		if msg == "" {
			t.Errorf("AAErrorMessages[%s] has empty message", code)
		}
		if code == "" {
			t.Error("AAErrorMessages has empty key")
		}
	}
}

func TestAAErrorMessages_Count(t *testing.T) {
	// Should have exactly 20 entries based on the source
	expectedCount := 20
	if len(AAErrorMessages) != expectedCount {
		t.Errorf("AAErrorMessages count = %d, want %d", len(AAErrorMessages), expectedCount)
	}
}

func TestAddressConstants_NotEmpty(t *testing.T) {
	if EntryPoint07Address == "" {
		t.Error("EntryPoint07Address should not be empty")
	}
	if Safe4337ModuleAddress == "" {
		t.Error("Safe4337ModuleAddress should not be empty")
	}
	if SafeWebAuthnSharedSigner == "" {
		t.Error("SafeWebAuthnSharedSigner should not be empty")
	}
	if FCLP256Verifier == "" {
		t.Error("FCLP256Verifier should not be empty")
	}
	if P256OwnerFactory == "" {
		t.Error("P256OwnerFactory should not be empty")
	}
	if WebAuthnSignerFactory == "" {
		t.Error("WebAuthnSignerFactory should not be empty")
	}
}
